File: leetcode/problems/language-specific/go/22.generate-parentheses.go
/*
 * @lc app=leetcode id=22 lang=golang
 *
 * [22] Generate Parentheses
 */

// @lc code=start
func generateParenthesis(n int) []string {
	res := []string{}
	generateParenthesisHelper(n, &res, "", 0, 0)
	return res
}

func generateParenthesisHelper(n int, res *[]string, cur string, open, close int) {
	if len(cur) == n*2 {
		*res = append(*res, cur)
		return
	}

	if open < n {
		generateParenthesisHelper(n, res, cur+"(", open+1, close)
	}

	if close < open {
		generateParenthesisHelper(n, res, cur+")", open, close+1)
	}
}

// @lc code=end

File: leetcode/problems/language-specific/python/22.generate-parentheses.py
#
# @lc app=leetcode id=22 lang=python3
#
# [22] Generate Parentheses
#

# @lc code=start

# * Recursive Backtracking Solution | O((4^n)/√n) Time | O((4^n)/√n) Space
# * n -> Given n pairs

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []

        def generateParenthesisHelper(stack=[], open=0, close=0):
            if open == close == n:
                result.append(''.join(stack))
                return

            if open < n:
                stack.append('(')
                generateParenthesisHelper(stack, open+1, close)
                stack.pop()

            if close < open:
                stack.append(')')
                generateParenthesisHelper(stack, open, close+1)
                stack.pop()

        generateParenthesisHelper()
        return result

# @lc code=end
